### PR TITLE
Fail gracefully if the application is not allowed to write files #73

### LINKED ISSUE #73: Fail gracefully if the application is not allowed to write files
If the application is run in a folder in which it has no permissions to create files, the application should inform the user about the problem and fail gracefully.

### PR DESCRIPTION
**Pull request by ernestlim8**
Fixes #73 

After a command, AB3 displays an error message and exits after three
seconds.

Automatically exiting the application is more user-friendly because the
user is unable to use the application purposefully with the file write
error.

Let's check for write permissions, then display the error message to the user showing
the file that has no permissions. The application will automatically
exit gracefully after three seconds.

### CODE REVIEW

**Comment by ernestlim8** [comment]
Fixes #73

**Comment by damithc** [comment]
> Fixes #73

This should be in the first post for GitHub to close the issue when this PR is merged.

File: src/main/java/seedu/address/ui/MainWindow.java [hunk 1]
```diff
@@ -188,9 +192,37 @@ private CommandResult executeCommand(String commandText) throws CommandException
...
+                    logger.info("Exception occurred while waiting for thread");
+                } finally {
+                    // Execute handleExit on the JavaFx Thread
+                    Platform.runLater(() -> handleExit());
+                }
+                return null;
+            }
+        };
+
+        new Thread(accessDeniedTask).start();
```
Comment (JinHao-L) [inline src/main/java/seedu/address/ui/MainWindow.java]: I found another alternative solution which is the [`PauseTransition`](https://docs.oracle.com/javase/8/javafx/api/javafx/animation/PauseTransition.html) class in the JavaFX library.

The following code produces the same effect and similarly does not freeze the UI. 
```
PauseTransition pause = new PauseTransition(Duration.seconds(3));
pause.setOnFinished(event -> handleExit());
pause.play();
```

I am not sure if this is more efficient but I feel that this solution might be more reader-friendly. 
What do you all think?

File: src/main/java/seedu/address/ui/MainWindow.java [hunk 2]
```diff
@@ -188,9 +192,37 @@ private CommandResult executeCommand(String commandText) throws CommandException
...
+                logger.info("Invalid command: " + commandText);
+            }
+
             throw e;
         }
     }
+
+    private void handleAccessDeniedException(Exception e) {
+        logger.info("Insufficient Permissions for file");
+        resultDisplay.setFeedbackToUser(e.getMessage());
```
Comment (JinHao-L) [inline src/main/java/seedu/address/ui/MainWindow.java]: Currently, the feedback to the user seems like an unchecked exception:
> Could not save data to file: java.nio.file.AccessDeniedException: data\addressbook.json

Maybe it would be better if we rephrase it to `"Could not save data to file: Access Denied"` or `"Could not save data to file: Insufficient permission"`.

**Review by JinHao-L — COMMENTED** [review]
Other than the comments below, the rest looks good! 👍

File: src/main/java/seedu/address/ui/MainWindow.java (hunk 1 above)
Comment (ernestlim8) [reply src/main/java/seedu/address/ui/MainWindow.java]: Agreed, this solution seems more user friendly. It works on my end as well.

File: src/main/java/seedu/address/ui/MainWindow.java (hunk 2 above)
Comment (ernestlim8) [reply src/main/java/seedu/address/ui/MainWindow.java]: I have updated the error message to be more readable for the user as per your suggestion.

`Could not save data to file: Insufficient permission for <filepath>`
The additional information of the file path allows the user to know which file path has insufficient permissions.

**Review by damithc — COMMENTED** [review]
Thanks for the PR @ernestlim8 
I think it is enough to show a clear error message and stop operations. 
There is no need to exit automatically. Auto-exiting will make test automation harder (although we are not doing auto-testing of the GUI in AB3, it is done in AB4). Even in manual testing, it will make it harder for students to take screenshots if the error happens.

**Comment by ernestlim8** [comment]
@damithc After commenting out the Pause Transition code, if there are insufficient file permissions, AB3 will display the error message for all valid commands. Is this behaviour sufficient for stopping operations?

**Comment by damithc** [comment]
> @damithc After commenting out the Pause Transition code, if there are insufficient file permissions, AB3 will display the error message for all valid commands. Is this behaviour sufficient for stopping operations?

Sorry @ernestlim8 , overlooked this until now due to start-of-semester rush. Yes, I think that would work. Is the current code doing just that? I'm asking because I saw some unusual import statements.

**Comment by damithc** [comment]
@se-edu/tech-team-level1 for your review

**Comment by damithc** [comment]
Sorry @ernestlim8 I think this PR will progress very slowly as this is the semester time and this change is not urgent.

File: src/main/java/seedu/address/ui/MainWindow.java [hunk 3]
```diff
@@ -188,9 +190,19 @@ private CommandResult executeCommand(String commandText) throws CommandException
...
             return commandResult;
         } catch (CommandException | ParseException e) {
-            logger.info("Invalid command: " + commandText);
-            resultDisplay.setFeedbackToUser(e.getMessage());
+            if (e.getCause() instanceof AccessDeniedException) {
+                handleAccessDeniedException(e);
+            } else {
+                resultDisplay.setFeedbackToUser(e.getMessage());
+                logger.info("Invalid command: " + commandText);
+            }
```
Comment (damithc) [inline src/main/java/seedu/address/ui/MainWindow.java]: See if you can get rid of this condition (checking of object type should be avoided if possible). That is, if the correct error message was put into the exception in the first place, you don't need a different behaviors here. That means we only need to change the place where the error is detected.

**Review by damithc — CHANGES_REQUESTED** [review]

File: src/main/java/seedu/address/ui/MainWindow.java [hunk 4]
```diff
@@ -188,8 +189,8 @@ private CommandResult executeCommand(String commandText) throws CommandException
 
             return commandResult;
         } catch (CommandException | ParseException e) {
-            logger.info("Invalid command: " + commandText);
             resultDisplay.setFeedbackToUser(e.getMessage());
+            logger.info("Invalid command: " + commandText);
```
Comment (damithc) [inline src/main/java/seedu/address/ui/MainWindow.java]: Any reason for this change?

File: src/main/java/seedu/address/ui/MainWindow.java [hunk 5]
```diff
@@ -156,6 +156,7 @@ void show() {
      */
     @FXML
     private void handleExit() {
+        logger.info("Calling exit");
```
Comment (damithc) [inline src/main/java/seedu/address/ui/MainWindow.java]: Seems this change is not related to this PR?

**Review by damithc — COMMENTED** [review]
Sorry for getting back so late @ernestlim8 
Some minor comments added.

**Comment by damithc** [comment]
@ernestlim8 shall we finish this PR?

File: src/main/java/seedu/address/logic/LogicManager.java [hunk 6]
```diff
@@ -48,7 +48,8 @@ public CommandResult execute(String commandText) throws CommandException, ParseE
         try {
             storage.saveAddressBook(model.getAddressBook());
         } catch (IOException ioe) {
-            throw new CommandException(FILE_OPS_ERROR_MESSAGE + ioe, ioe);
+            String message = ioe.getMessage();
+            throw new CommandException(String.format(FILE_OPS_ERROR_MESSAGE, message), ioe);
```
Comment (damithc) [inline src/main/java/seedu/address/logic/LogicManager.java:52]: Looks like the code assumes every IOException is caused by _insufficient permissions_? I don't think that is a valid assumption.

File: src/main/java/seedu/address/ui/MainWindow.java [hunk 7]
```diff
@@ -188,7 +188,6 @@ private CommandResult executeCommand(String commandText) throws CommandException
 
             return commandResult;
         } catch (CommandException | ParseException e) {
-            logger.info("Invalid command: " + commandText);
```
Comment (damithc) [inline src/main/java/seedu/address/ui/MainWindow.java:191]: What if the command is actually invalid?

**Review by damithc — COMMENTED** [review]
I have some concerns about the logic of this fix. Happy to be proven wrong though.

File: src/main/java/seedu/address/ui/MainWindow.java (hunk 7 above)
Comment (ernestlim8) [reply src/main/java/seedu/address/ui/MainWindow.java:191]: @damithc Seems like this log statement was already existing before this PR, its possible to split the 2 different exceptions and handle them separately, if the desired behaviour is to differentiate between invalid command and parse exception in the logs

File: src/main/java/seedu/address/logic/LogicManager.java (hunk 6 above)
Comment (ernestlim8) [reply src/main/java/seedu/address/logic/LogicManager.java:52]: @damithc Agreed, I will display the insufficient permissions when catching the AccessDeniedException instead, and use a more generalised message for other possible I/O Exceptions.

**Comment by damithc** [comment]
To be continued by #172 
Thanks for this PR @ernestlim8
